### DISCUSSION TITLE
Mastodon on a different server than the NGINX server

### DISCUSSION THREAD

**Discussion by Ahwxorg** [mastodon/mastodon#27379] [General]
Hey there, I have the following setup:

proxmox machine:
* `vm 1`, a vm with NGINX Proxy Manager

another proxmox machine:
* `vm 2`, Mastodon vm

I want to let `vm 1` do the proxying for `vm 2` with Mastodon on it. How do I do this? I currently have in NPM:

![image](https://github.com/mastodon/mastodon/assets/87855602/6919646f-4664-4cf6-91e0-fc11b7e6c6d5)
![image](https://github.com/mastodon/mastodon/assets/87855602/3ba79515-4479-43b0-8d2e-80020a690699)
![image](https://github.com/mastodon/mastodon/assets/87855602/2fc7f5ad-1cd5-4046-a912-bb1b4ff3c52d)
![image](https://github.com/mastodon/mastodon/assets/87855602/39af1f9a-3529-4a17-90dd-d249f7a82730)

Custom NGINX configuration is:

```ini
gzip on;
gzip_disable "msie6";
gzip_vary on;
gzip_proxied any;
gzip_comp_level 6;
gzip_buffers 16 8k;
gzip_http_version 1.1;
gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript;

proxy_request_buffering off;
proxy_buffering off;
proxy_redirect off;
tcp_nodelay on;
```

**Comment by jdoubleu** [comment]
what response do you get when you
```
curl -I 10.10.0.235:3000
```

from `vm1`?

**Comment by Ahwxorg** [comment]
`vm1` does not run Mastodon. It returns:
```sh
root@pve-dl380p-1-nginxproxymanager:~# curl -I 127.0.0.1:3000
HTTP/1.1 200 OK
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Pragma: no-cache
Expires: 0
Content-Type: application/json; charset=utf-8
Content-Length: 61
ETag: W/"3d-gRCNZEPviTX3/BresR/AlzDxfGo"
Vary: Accept-Encoding
Date: Thu, 12 Oct 2023 17:37:04 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

`vm2` returns the following:
```sh
root@pve-dl380p-2-ahwxorg:~# curl -I http://localhost:3000
curl: (7) Failed to connect to localhost port 3000 after 0 ms: Couldn't connect to server
```

**Reply by jdoubleu** [reply]
sorry, meant to write `10.10.0.235`.

can you reach vm2 from vm1?

**Reply by Ahwxorg** [reply]
![image](https://github.com/mastodon/mastodon/assets/87855602/ee425b66-1116-4b39-a1d4-9a9ed5c2ca00)

**Reply by nightpool** [reply]
```
root@pve-dl380p-2-ahwxorg:~# curl -I http://localhost:3000
curl: (7) Failed to connect to localhost port 3000 after 0 ms: Couldn't connect to server
```

Sounds like maybe the Mastodon server isn't running then, or is running on a different port than you expect it to be running on? You configured Nginx to point to vm2:3000, but this seems to indicate that vm2:3000 doesn't have Mastodon running (at least at the time you ran the curl command)

**Reply by jdoubleu** [reply]
> Sounds like maybe the Mastodon server isn't running then

That's `vm1`, where the proxy is, not the mastodon instance.

What I'm trying to check is, whether the proxy can even reach mastodon. And not just through ICMP but by HTTP via port 3000.

Please check whether you can reach `10.10.0.235:3000` from your proxy.

**Reply by Ahwxorg** [reply]
![image](https://github.com/mastodon/mastodon/assets/87855602/d93b8c1f-a7e1-4574-b4fb-15ab04f5edc1)

![image](https://github.com/mastodon/mastodon/assets/87855602/eb469074-315d-486e-a140-bf45087a6fd5)

The VMs/servers can reach eachother. They are on the same VLAN and subnet. What can I do? I think Mastodon hosts itself on localhost:3000 and does not want to be discovered on the interface that is accessible from the LAN.

**Comment by BossMAN559** [comment]
have you tried to add 
ALLOWED_PRIVATE_ADDRESSES=10.10.0.6, 10.10.0.235 
to your mastodon config?

**Reply by nightpool** [reply]
ALLOWED_PRIVATE_ADDRESSES is only for requests *Mastodon* makes to other servers, not for requests other servers (like nginx) are making to Mastodon

**Comment by nightpool** [comment]
There isn't any information in your question about what part of your setup isn't currently working. The setup you have seems to make sense to me from a conceptual level, so it sounds like you just have a simple configuration issue or something else that isn't set up correctly. I would recommend using the actual `nginx.conf` template shipped with mastodon, and then modifying the upstreams manually, rather than trying to replicate it within a web UI like nginx proxy manager.

**Reply by Ahwxorg** [reply]
Would be the issue indeed, but sadly I can't even reach Mastodon from the proxy server. After that I think it'd be pretty easy...
